packages/oauth/src/index.ts

```diff
@@ -137,6 +195,52 @@ export class InstallProvider {
     }
   }
 
+  /**
+   * checkTokenExpiry determines access tokens' eligibility for refresh.
+   *
+   * The return value is an Array of expired or soon-to-expire access tokens.
+   */
+  private checkTokenExpiry(authResult: AuthorizeResult, currentUTCSec: number): string[] {
+    const tokensToRefresh: string[] = [];
+    const EXPIRY_WINDOW: number = 7200; // 2 hours
+
+    if (authResult.botRefreshToken !== undefined && authResult.botTokenExpiresAt !== undefined) {
+      const botTokenExpiresIn = authResult.botTokenExpiresAt - currentUTCSec;
+      if (botTokenExpiresIn <= EXPIRY_WINDOW) {
+        tokensToRefresh.push(authResult.botRefreshToken);
+      }
+    }
+
+    if (authResult.userRefreshToken !== undefined && authResult.userTokenExpiresAt !== undefined) {
+      const userTokenExpiresIn = authResult.userTokenExpiresAt - currentUTCSec;
+      if (userTokenExpiresIn <= EXPIRY_WINDOW) {
+        tokensToRefresh.push(authResult.userRefreshToken);
+      }
+    }
+
+    return tokensToRefresh;
+  }
+
+  /**
+   * refreshExpiringTokens refreshes expired access tokens using the `oauth.v2.access` endpoint.
+   *
+   * The return value is an Array of Promises made up of the resolution of each token refresh attempt.
+   */
+  private async refreshExpiringTokens(tokensToRefresh: string[]): Promise<OAuthV2ExchangeResponse[]> {
+    const client = new WebClient(undefined, this.clientOptions);
+
+    const refreshPromises = tokensToRefresh.map(async (refreshToken) => {
+      return await client.oauth.v2.access({
+        client_id: this.clientId,
+        client_secret: this.clientSecret,
+        grant_type: 'refresh_token',
+        refresh_token: refreshToken,
+      }).catch(e => e) as OAuthV2ExchangeResponse;
+    });
+
+    return Promise.all(refreshPromises);
```

Comment:
I spent a good amount of time trying to best solution this, but would love input. 

With our current tsconfig, we don't have the ability to use `Promise.allSettled` (we use `es2017` and would need `es2020`), which is what we want in this case (to wait until all Promises are settled, despite outcome of each).

`Promise.all` will fail fast if a rejection occurs, and I think under normal circumstances that would be handled fine with the additional `.catch(e => e)` that's included for each individual Promise, but in the case of a failure, the WebAPI actually throws an error before that catch can happen (I _think_ that's what I've been observing while testing. Feel free to challenge that).

Reply:
As we dropped Node 10 support, perhaps, `Promise.allSettled` should be available in all supported Node runtimes.